internal/build/imgsrc/ensure_builder.go

```diff
@@ -0,0 +1,325 @@
+		flaps, err := flapsutil.NewClientWithOptions(ctx, flaps.NewClientOpts{
+			AppName: builderApp.Name,
+			// TOOD(billy) make a utility function for App -> AppCompact
+			AppCompact: &fly.AppCompact{
+				ID:       builderApp.ID,
+				Name:     builderApp.Name,
+				Status:   builderApp.Status,
+				Deployed: builderApp.Deployed,
+				Hostname: builderApp.Hostname,
+				AppURL:   builderApp.AppURL,
+				Organization: &fly.OrganizationBasic{
+					ID:       builderApp.Organization.ID,
+					Name:     builderApp.Organization.Name,
+					Slug:     builderApp.Organization.Slug,
+					RawSlug:  builderApp.Organization.RawSlug,
+					PaidPlan: builderApp.Organization.PaidPlan,
+				},
+				PlatformVersion: builderApp.PlatformVersion,
+				PostgresAppRole: builderApp.PostgresAppRole,
+			},
+			OrgSlug: builderApp.Organization.Slug,
+		})
+		if err != nil {
+			tracing.RecordError(span, err, "error creating flaps client")
+			return nil, nil, err
+		}
+		ctx = flapsutil.NewContextWithClient(ctx, flaps)
+	}
+
+	builderMachine, err := validateBuilder(ctx, builderApp)
+	if err == nil {
+		span.AddEvent("builder app already exists and is valid")
+		return builderMachine, builderApp, nil
+	}
+
+	var validateBuilderErr ValidateBuilderError
+	if !errors.As(err, &validateBuilderErr) {
+		return nil, nil, err
+	}
+
+	if validateBuilderErr != NoBuilderApp {
+		span.AddEvent(fmt.Sprintf("deleting existing invalid builder due to %s", validateBuilderErr))
+		client := flyutil.ClientFromContext(ctx)
+		err := client.DeleteApp(ctx, builderApp.Name)
+		if err != nil {
+			tracing.RecordError(span, err, "error deleting invalid builder app")
+			return nil, nil, err
+		}
+	}
+
+	builderName := "fly-builder-" + haikunator.Haikunator().Build()
+	// we want to lauch the machine to the builder
+	flapsClient, err := flapsutil.NewClientWithOptions(ctx, flaps.NewClientOpts{
+		AppName: builderName,
+		OrgSlug: org.Slug,
+	})
+	if err != nil {
+		tracing.RecordError(span, err, "error creating flaps client")
+		return nil, nil, err
+	}
+	ctx = flapsutil.NewContextWithClient(ctx, flapsClient)
+	app, machine, err := createBuilder(ctx, org, region, builderName)
+	return machine, app, err
+}
+
+type ValidateBuilderError int
+
+func (e ValidateBuilderError) Error() string {
+	switch e {
+	case NoBuilderApp:
+		return "no builder app"
+	case NoBuilderVolume:
+		return "no builder volume"
+	case InvalidMachineCount:
+		return "invalid machine count"
+	default:
+		return "unknown error validating builder"
+	}
+}
+
+const (
+	NoBuilderApp ValidateBuilderError = iota
+	NoBuilderVolume
+	InvalidMachineCount
+)
+
+func validateBuilder(ctx context.Context, app *fly.App) (*fly.Machine, error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "validate_builder")
+	defer span.End()
+
+	if app == nil {
+		tracing.RecordError(span, NoBuilderApp, "no builder app")
+		return nil, NoBuilderApp
+	}
+
+	flapsClient := flapsutil.ClientFromContext(ctx)
+
+	var volumes []fly.Volume
+	numRetries := 0
```

Comment:
We use a "backoff" library elsewhere for retries. Maybe that's better to use than rolling your own?

Reply:
the only issue with the backoff library is that i need to exit early if we don't get a server error, instead of retrying multiple times (for example, if we get a 404, it's likely a user error)